redis/src/tls.rs

```diff
@@ -0,0 +1,212 @@
+use std::io::{BufRead, Error, ErrorKind as IOErrorKind};
+
+use rustls::{Certificate, OwnedTrustAnchor, PrivateKey, RootCertStore};
+use rustls_pemfile::{certs, read_one};
+
+use crate::{
+    Client, ConnectionAddr, ConnectionInfo, ErrorKind, RedisConnectionInfo, RedisError, RedisResult,
+};
+
+/// The builder takes all parameters necessary to create a TLS connection.
+///
+/// In addition to a `server` and a `port`, it must be provided with
+/// binary streams corresponding to PEM file's contents for:
+/// - CA's certificate
+/// - one client certificate and
+/// - client's key
+///
+/// Example usage:
+///
+/// ```rust,no_run
+/// use redis::{self, build_with_tls};
+/// use redis::AsyncCommands;
+///
+/// use std::fs::File;
+/// use std::io::BufReader;
+///
+/// async fn do_redis_code(server: &str, port: u16, ca_file: &str, cert_file: &str, key_file: &str) -> redis::RedisResult<()> {
+///     let cert_file = File::open(cert_file).expect("cannot open private cert file");
+///     let mut reader_cert = BufReader::new(cert_file);
+///
+///     let key_file = File::open(key_file).expect("Cannot open private key file");
+///     let mut reader_key = BufReader::new(key_file);
+///
+///     let ca_file = File::open(ca_file).expect("Cannot open CA cert file");
+///     let mut reader_ca = BufReader::new(ca_file);
+///
+///     let client = build_with_tls(server, port, &mut reader_cert, &mut reader_key, &mut reader_ca).expect("Unable to build client");
+///
+///     let connection_info = client.get_connection_info();
+///
+///     println!(">>> connection info: {connection_info:?}");
+///
+///     let mut con = client.get_async_connection().await?;
+///
+///     con.set("key1", b"foo").await?;
+///
+///     redis::cmd("SET")
+///         .arg(&["key2", "bar"])
+///         .query_async(&mut con)
+///         .await?;
+///
+///     let result = redis::cmd("MGET")
+///         .arg(&["key1", "key2"])
+///         .query_async(&mut con)
+///         .await;
+///     assert_eq!(result, Ok(("foo".to_string(), b"bar".to_vec())));
+///     println!("Result from MGET: {result:?}");
+///
+///     Ok(())
+/// }
+/// ```
+pub fn build_with_tls(
+    server: impl ToString,
+    port: u16,
+    mut client_cert: &mut dyn BufRead,
+    mut client_key: &mut dyn BufRead,
+    mut root_cert: &mut dyn BufRead,
+) -> RedisResult<Client> {
+    let mut tls_builder = TlsClientBuilder::new(server, port);
+
+    tls_builder.client_cert(&mut client_cert)?;
+
+    tls_builder.client_key(&mut client_key)?;
+
+    tls_builder.root_cert(&mut root_cert)?;
+
+    tls_builder.build()
+}
+
+#[derive(Debug, Clone)]
+pub struct TlsConnParams {
+    pub(crate) client_key: PrivateKey,
+    pub(crate) client_cert: Vec<Certificate>,
+    pub(crate) root_cert: RootCertStore,
+}
+
+/// The TLS client builder type
+struct TlsClientBuilder {
```

Comment:
why keep this struct? it adds almost 100 lines to the file. AFAIS TlsClientBuilder ::build can be inlined into  build_with_tls, or kept as a free function, and the builder can be removed.

Reply:
dang, you are right

Reply:
thank you :)